src/code42cli/click_ext/types.py

```diff
@@ -16,3 +20,99 @@ def convert(self, value, param, ctx):
             return file.read()
         else:
             return value
+
+
+class MagicDate(click.ParamType):
+    """Declares a parameter to be a 'magic' date string. Accepts an optional `round` argument
+    which can be a function that takes a datetime and returns it rounded appropriately. This allows
+    imprecise "day" input values (2020-01-01, 3d) to be rounded to the start or end of the day
+    if needed. Accepts the following values as user input:
+
+    timestamp formats:
+        yyyy-MM-dd
+        yyyy-MM-dd HH
+        yyyy-MM-dd HH:MM
+        yyyy-MM-dd HH:MM:SS
+
+    short-string (day, hour, min) formats:
+        30d
+        24h
+        15m
+
+    and converts them to datetime objects.
+    """
+
+    TIMESTAMP_REGEX = re.compile(r"(\d{4}-\d{2}-\d{2})\s*(.*)?")
+    MAGIC_TIME_REGEX = re.compile(r"(\d+)([dhm])$")
+    HELP_TEXT = (
+        "Accepts a date/time in yyyy-MM-dd (UTC) or yyyy-MM-dd HH:MM:SS "
+        "(UTC+24-hr time) format where the 'time' portion of the string "
+        "can be partial (e.g. '2020-01-01 12' or '2020-01-01 01:15') or "
+        "a 'short time' value representing days (30d), hours (24h) or "
+        "minutes (15m) from current time."
+    )
+
+    name = "magicdate"
+
+    def __init__(self, rounding_func=None):
+        self.round = rounding_func
+
+    def get_metavar(self, param):
+        return "[DATE|TIMESTAMP|SHORT_TIME]"
+
+    def __repr__(self):
+        return "MagicDate"
```

Comment:
Does this have a CLI use?  It seems to not offer much context, less than not defining a `__repr__` even because it is might easy to miscontrue with the repr of the str `MagicDate`

Reply:
repr determines some stuff that shows up in debuggers when hovering over stuff while you've hit breakpoint, which can be helpful. In this case though since what's returned is just the name of the class, I'm not sure it's needed.

Reply:
Ah, I see. If it makes debugging look cleaner, than it makes sense I suppose... The default repr just includes a mem address which may seem noisy

Reply:
I just copied the pattern from an existing click.ParamType subclass, I'm not exactly sure when __repr__ will be useful, but it can't hurt.